### PR TITLE
Identity refactor (final step)

### LINKED ISSUE #162: Split `identity` into signature specific crates
`identity-dsa` for EcDSA and Ed25519, and `identity-hsm` for HSM, and `identity-secpk256r1`, etc etc

### PR DESCRIPTION
**Pull request by hansl**
I think we can go further in abstractions and API design, but this is good enough to get us a good way.

This splits `many-identity` into 4; 
- `many-identity` itself contains the Address implementation, and the Identity and Verifier traits.
- `many-identity-dsa` contains both `ed25519` and `ecdsa` methods and identities, as well as verifiers to validate the signatures.
- `many-identity-hsm` contains the identity to sign messages with an HSM. The `ecdsa` verifiers should be used to validate the signature.
- `many-identity-webauthn` contains the verifier to validate messages that are signed using our custom WebAuthn method.

This still passes all the tests (including HSM). I am still investigating code coverage and hope to add more, but the review can be started.

----

Fixes #162

### CODE REVIEW

**Comment by hansl** [comment]
Adding @Denommus as he might have some insights into the design patterns used here and how to improve them.

File: src/many-identity/src/identity.rs
Comment (hansl) [reply src/many-identity/src/identity.rs:106]: That's great feedback, and you're right. I'll remove it and implement that.

File: src/many-identity/src/identity.rs
Comment (Denommus) [inline src/many-identity/src/identity.rs:63]: I don't understand the point of this impl. `AsRef<T>` is implemented for `Box<T>`, so every `Box<dyn T>` can be borrowed as a `&dyn self` ref and call the `&self` methods on it. Doesn't the code compile without these redundant impls? Can you show me it breaking?

File: src/many-identity/src/identity.rs
Comment (Denommus) [inline src/many-identity/src/identity.rs:77]: Same for this impl, it's redundant with ` AsRef<T> for Box<T>`

**Review by Denommus — CHANGES_REQUESTED** [review]
The PR in general looks good, but some implementations add unnecessary verbosity and are redundant with what Rust already offers.

File: src/many-identity/src/identity.rs
Comment (hansl) [reply src/many-identity/src/identity.rs:63]: Nope.

image

**Review by Denommus — APPROVED** [review]

File: src/many-identity-hsm/src/lib.rs
Comment (fmorency) [inline src/many-identity-hsm/src/lib.rs:314]: You removed the ability to perform hardware verification of messages. Isn't hardware offloading something we wanted for, e.g., servers?

File: src/many-identity/src/identity.rs
Comment (fmorency) [inline src/many-identity/src/identity.rs:135]: Might be worth adding a `trace!`.

File: src/many-identity/src/identity.rs
Comment (fmorency) [inline src/many-identity/src/identity.rs:119]: Noob question: can we use const generic instead of this?

File: src/many-modules/src/_9_account.rs
Comment (fmorency) [inline src/many-modules/src/_9_account.rs:476]: `identity(0)`?

File: src/many-modules/src/_9_account.rs
Comment (fmorency) [inline src/many-modules/src/_9_account.rs:679]: `identity(0)`?

File: src/many-modules/src/_9_account.rs
Comment (fmorency) [inline src/many-modules/src/_9_account.rs:689]: `identity(1)`

File: src/many-modules/src/_9_account.rs
Comment (fmorency) [inline src/many-modules/src/_9_account.rs:693]: `identity(1)`

File: src/many-modules/src/_9_account.rs
Comment (fmorency) [inline src/many-modules/src/_9_account.rs:698]: `identity(1)`

File: src/many-modules/src/_9_account.rs
Comment (fmorency) [inline src/many-modules/src/_9_account.rs:703]: `identity(1)`

File: src/many-server/proptest-regressions/server.txt
Comment (fmorency) [inline src/many-server/proptest-regressions/server.txt:7]: Remove?

File: src/many-protocol/src/lib.rs
Comment (fmorency) [inline src/many-protocol/src/lib.rs:43]: I'll comment on it only once, but this also applies to `many-server`.

What does this do, and how is it used? 

I don't get it.

File: src/many-identity-dsa/src/impls/ed25519.rs
Comment (fmorency) [inline src/many-identity-dsa/src/impls/ed25519.rs:163]: Why is `Arc` needed?

**Review by fmorency — CHANGES_REQUESTED** [review]
First iteration. Thanks!

**Comment by fmorency** [comment]
Please also update the crates diagram & description in the README file.

File: src/many-error/src/error.rs
Comment (hansl) [reply src/many-error/src/error.rs:152]: We don't (yet) have any guarantees for error codes, and I'd like to keep the grouped.

We need a full review of all errors. There's a lot of `ManyError::unknown()` in this PR specifically because we need to catalog them all and review our whole error strategy.

File: src/many-identity/src/identity.rs
Comment (hansl) [reply src/many-identity/src/identity.rs:13]: Fair. I renamed it to `verify_1`.

File: src/many-identity-dsa/src/impls/ed25519.rs
Comment (hansl) [reply src/many-identity-dsa/src/impls/ed25519.rs:60]: It's not the same `public_key()` being called, but I'll see what I can do.

File: src/many-identity-dsa/src/lib.rs
Comment (hansl) [reply src/many-identity-dsa/src/lib.rs:153]: Yeah. This will complicate the code later if we add more EdDSA curves, for example. Unless we can see this is a bottleneck I'd like to defer whether or not a key is supported to its implementation, for now.

File: src/many-identity-dsa/src/impls/ed25519.rs
Comment (hansl) [reply src/many-identity-dsa/src/impls/ed25519.rs:227]: Not yet, but I forgot to add it and I was thinking how it's not much code. I can remove it.

File: src/many-identity-hsm/src/lib.rs
Comment (hansl) [reply src/many-identity-hsm/src/lib.rs:314]: Right. Let me see if I can re-add it.

File: src/many-identity/src/address.rs
Comment (hansl) [reply src/many-identity/src/address.rs:186]: Unsafe here just means "watch out", which is fair IMO.

File: src/many-identity/src/identity.rs
Comment (hansl) [reply src/many-identity/src/identity.rs:42]: On the error or the success?

Comment (hansl) [reply src/many-identity/src/identity.rs:42]: Actually removed this. Don't use an identity to verify messages.

File: src/many-identity/src/identity.rs
Comment (hansl) [reply src/many-identity/src/identity.rs:119]: Not for tuples.

File: src/many-protocol/src/lib.rs
Comment (hansl) [reply src/many-protocol/src/lib.rs:43]: I'll remove it, that's a bit premature.

File: src/many-identity-dsa/src/lib.rs
Comment (fmorency) [reply src/many-identity-dsa/src/lib.rs:153]: For the sake of posterity, the `CoseKey` structure should contain all this information already (key type, algorithm, curve type, etc.).

See https://datatracker.ietf.org/doc/html/rfc8152#page-41 for example.

File: src/many-protocol/src/lib.rs
Comment (hansl) [reply src/many-protocol/src/lib.rs:43]: Actually, I'll keep it there. It is the bit of code that takes a message and return the origin address of the message.

This is (currently) the hash of the public key (or the `kid` too), but it could be different. It's also something that requires more logic than reading a single field because it should match the signature.

This was done by the Verifier before (or the code that was doing the verification), but there are cases that become very complex and shouldn't be left to the verifier itself (like delegation).

I just now noticed that this logic was always missing. I'll re-add it to make sure we don't introduce a bug.

File: src/many-identity-dsa/src/impls/ed25519.rs
Comment (hansl) [reply src/many-identity-dsa/src/impls/ed25519.rs:163]: `Keypair` does not have `Clone`.

Comment (hansl) [reply src/many-identity-dsa/src/impls/ed25519.rs:163]: I removed `Clone` from all keys. We can reintroduce it later if we need to.

**Review by fmorency — CHANGES_REQUESTED** [review]
Thanks! Some additional comments.

File: README.md
Comment (hansl) [reply README.md:37]: Created https://github.com/liftedinit/many-rs/issues/188

File: src/many-identity-dsa/src/impls.rs
Comment (hansl) [reply src/many-identity-dsa/src/impls.rs:62]: `ecdsa` doesn't have a curve.

File: src/many-identity/src/cose.rs
Comment (hansl) [reply src/many-identity/src/cose.rs:12]: Sure, that's sensible.

File: src/many-identity-dsa/src/impls.rs
Comment (fmorency) [reply src/many-identity-dsa/src/impls.rs:62]: Right. 

We should probably check it, so we have consistent behavior.

Comment (hansl) [reply src/many-identity-dsa/src/impls.rs]: https://github.com/liftedinit/many-rs/issues/189

File: src/many-identity-dsa/src/lib.rs
Comment (hansl) [reply src/many-identity-dsa/src/lib.rs:13]: Done. Let's see what the coverage looks like now.

Comment (hansl) [reply src/many-identity-dsa/src/lib.rs:13]: It's now at 61.98% so I'll resolve this.

File: src/many-identity-hsm/src/lib.rs
Comment (fmorency) [reply src/many-identity-hsm/src/lib.rs:314]: I don't see this resolved, but I won't block the merge. You can create a new issue and add it to another PR.

**Review by fmorency — APPROVED** [review]
Approving, assuming all open discussions are resolved.

File: src/many-identity-hsm/src/lib.rs
Comment (hansl) [reply src/many-identity-hsm/src/lib.rs:314]: This isn't used anywhere yet. We should file an issue and prioritize it properly.

File: src/many-protocol/src/lib.rs
Comment (hansl) [reply src/many-protocol/src/lib.rs]: `verify_1` now returns the address that matched the verification. Additionaly resolution steps might be required.

File: src/many-identity-hsm/src/lib.rs
Comment (hansl) [reply src/many-identity-hsm/src/lib.rs:314]: https://github.com/liftedinit/many-rs/issues/190

**Pull request merged**
